### PR TITLE
Run test plans on Docker Swarm (AWS)

### PR DESCRIPTION
**Pull request by raulk**
This PR builds on #112. It introduces:

* a `cluster:swarm` runner that takes a docker image of a test plan, and runs a test case within a Docker swarm cluster on AWS.
* `docker:go` builder can now push to Docker registries, particularly AWS ECR registries. It does the `aws ecr get-login` dance and all!
* an environment configuration file for things like AWS secrets, docker endpoints, etc. This is what mine looks like (eliding sensitive info):

```toml
["aws"]
access_key_id = "access key id"
secret_access_key = "secret access key"
region = "us-west-2"

[run_strategies."cluster:swarm"]
docker_endpoint = "tcp://localhost:4545"
docker_tls = false
docker_tls_ca_cert_path = "/"
docker_tls_cert_path = "/"
docker_tls_key_path = "/"
```

  I'm hitting the remote Docker Swarm manager endpoint by creating an SSH tunnel for port forwarding:

```sh
$ ssh -nNT -L 4545:/var/run/docker.sock ubuntu@...compute.amazonaws.com
```

### CODE REVIEW

**Comment by daviddias** [comment]
Could you:

- [ ] Create a `secrets-example.toml` with the placeholder you pasted at https://github.com/ipfs/testground/pull/126#issue-337093730 and add to `.gitignore` a `secrets.toml` 
- [ ] Add to the README the note on how to create that file (i.e. `cp secrets-example.toml secrets.toml` and add the the user own secrets)

**Comment by raulk** [comment]
@daviddias good ideas!

File: pkg/engine/engine.go [hunk 1]
```diff
@@ -84,6 +88,16 @@ func NewEngine(cfg *EngineConfig) (*Engine, error) {
 	e.dirs.src = srcdir
 	e.dirs.work = workdir
 
+	// load the .env.toml file.
```
Comment (daviddias) [inline pkg/engine/engine.go]: The need for this .env.toml needs to be documented. See my suggestion at https://github.com/ipfs/testground/pull/126#issuecomment-550496063

File: pkg/engine/engine.go [hunk 2]
```diff
@@ -84,6 +88,16 @@ func NewEngine(cfg *EngineConfig) (*Engine, error) {
...
+	e.env = new(api.EnvConfig)
+	path := filepath.Join(srcdir, ".env.toml")
+	switch _, err := toml.DecodeFile(path, e.env); err.(type) {
+	case nil:
+		fmt.Println("loading env from", path)
+	case *os.PathError:
+		// TODO no .env.toml file located, maybe print a warning.
```
Comment (daviddias) [inline pkg/engine/engine.go]: Wanna do this TODO before the merge? It's super simple

File: pkg/runner/cluster_swarm.go [hunk 3]
```diff
@@ -1 +1,314 @@
...
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed while tailing logs: %w", err)
+	}
+
+	rpipe, wpipe := io.Pipe()
+	go func() {
+		_, err := stdcopy.StdCopy(wpipe, wpipe, rc)
+		_ = wpipe.CloseWithError(err)
+	}()
```
Comment (daviddias) [inline pkg/runner/cluster_swarm.go]: What's this?

Comment (raulk) [reply pkg/runner/cluster_swarm.go]: Adding comments.

File: pkg/engine/engine.go (hunk 2 above)
Comment (raulk) [reply pkg/engine/engine.go]: Sure.

File: pkg/engine/engine.go (hunk 1 above)
Comment (raulk) [reply pkg/engine/engine.go]: Done in https://github.com/ipfs/testground/pull/126/commits/a56b82c60d457a05060b71a882c8415edc63c195

File: pkg/build/golang/docker.go [hunk 4]
```diff
@@ -164,28 +182,86 @@ func (b *DockerGoBuilder) Build(opts *api.BuildInput) (*api.BuildOutput, error)
...
+	return out, nil
+}
+
+func (b *DockerGoBuilder) pushToRegistry(ctx context.Context, client *client.Client, in *api.BuildInput, out *api.BuildOutput) error {
+	cfg := in.BuildConfig.(*DockerGoBuilderConfig)
+
+	switch cfg.RegistryType {
+	case "aws":
+		goto AWS
+	}
```
[the remark this answers is not in the document]
Comment (raulk) [reply pkg/build/golang/docker.go]: True, this is a bug. Fixing!

File: cmd/run.go [hunk 5]
```diff
@@ -39,10 +40,13 @@ var RunCommand = cli.Command{
 			},
 			Usage: fmt.Sprintf("specifies the runner; options: %s", strings.Join(runners, ", ")),
 		},
+		cli.BoolFlag{
+			Name:  "no-build, nb",
+			Usage: "do not perform a build; requires --artifact-path option to be provided",
```
[the remark this answers is not in the document]
Comment (raulk) [reply cmd/run.go]: The `build` command generates an artifact and it spits out the artifact path. Then you can use `run` with `--no-build`, specifying the artifact path with `--artifact-path` to skip the build and use the artifact you build previously.

File: README.md [hunk 6]
```diff
@@ -128,6 +128,18 @@ NAME:
      --help, -h  show help
 ```
 
+### Setting an environment file
+
+Testground automatically loads an `.env.toml` file at root of your source
```
Comment (daviddias) [inline README.md]: Don't forget to add the .env.toml to `.gitignore` or many keys will be leaked by mistake :)

Comment (raulk) [reply README.md]: It was already there since the beginning 😛

Comment (daviddias) [reply README.md]: Ah! I missed it (was reviewing only the commit). Thanks ! :)

File: pkg/runner/cluster_swarm.go (hunk 3 above)
Comment (daviddias) [reply pkg/runner/cluster_swarm.go]: thank you!

File: cmd/run.go [hunk 7]
```diff
@@ -88,17 +94,28 @@ func runCommand(c *cli.Context) error {
...
+		return errors.New("artifact-path is required when skipping the build")
 	}
 
-	// Trigger the build job.
-	buildOut, err := _engine.DoBuild(comp[0], builderId, buildIn)
-	if err != nil {
-		return fmt.Errorf("error while building test plan: %w", err)
+	if !noBuild && artifactPath != "" {
+		logging.S().Warn("artifact-path will be ignored, as we're performing a build")
+	}
```
Comment (Stebalien) [inline cmd/run.go]: What about replacing both of these with "use-build"?

File: cmd/run.go [hunk 8]
```diff
@@ -88,17 +94,28 @@ func runCommand(c *cli.Context) error {
...
+
+	if !c.Bool("no-build") {
```
Comment (Stebalien) [inline cmd/run.go]: nit: `noBuild`?

File: pkg/aws/ecr.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+func (*ecrsvc) EnsureRepository(cfg api.AWSConfig, name string) (uri string, err error) {
+	creds := credentials.NewStaticCredentials(cfg.AccessKeyID, cfg.SecretAccessKey, "")
+	config := aws.NewConfig().WithRegion(cfg.Region).WithCredentials(creds)
+	sess, err := session.NewSession(config)
+	if err != nil {
+		return "", err
+	}
+
+	svc := ecr.New(sess)
+	svc.AddDebugHandlers()
```
Comment (Stebalien) [inline pkg/aws/ecr.go]: nit: Helper?

File: pkg/build/golang/docker.go [hunk 10]
```diff
@@ -164,28 +182,88 @@ func (b *DockerGoBuilder) Build(opts *api.BuildInput) (*api.BuildOutput, error)
...
+	cfg := in.BuildConfig.(*DockerGoBuilderConfig)
+
+	switch cfg.RegistryType {
+	case "aws":
+		goto AWS
+	default:
+		return fmt.Errorf("no registry type specified, or unrecognised value: %s", cfg.RegistryType)
+	}
+
+AWS:
```
Comment (Stebalien) [inline pkg/build/golang/docker.go]: 1. No need for this goto.
2. 

```go
if cfg.RegistryType != "aws" {
  return ...
}

File: pkg/engine/engine.go [hunk 11]
```diff
@@ -84,6 +88,16 @@ func NewEngine(cfg *EngineConfig) (*Engine, error) {
 	e.dirs.src = srcdir
 	e.dirs.work = workdir
 
+	// load the .env.toml file.
+	e.env = new(api.EnvConfig)
+	path := filepath.Join(srcdir, ".env.toml")
+	switch _, err := toml.DecodeFile(path, e.env); err.(type) {
+	case nil:
+		fmt.Println("loading env from", path)
+	case *os.PathError:
```
Comment (Stebalien) [inline pkg/engine/engine.go]: I assume you want `os.IsNotExist`?

File: pkg/engine/engine.go [hunk 12]
```diff
@@ -49,6 +52,7 @@ type Engine struct {
 		src  string
 		work string
 	}
+	env *api.EnvConfig
```
Comment (Stebalien) [inline pkg/engine/engine.go]: Probably doesn't need to be a pointer.

File: pkg/runner/cluster_swarm.go [hunk 13]
```diff
@@ -1 +1,318 @@
...
+	RemoveService: true,
+}
+
+// ClusterSwarmRunner is a runner that creates a Docker service to launch as
+// many replicated instances of a container as the run job indicates.
+type ClusterSwarmRunner struct{}
+
+// TODO runner option to keep containers alive instead of deleting them after
+// the test has run.
+func (*ClusterSwarmRunner) Run(input *api.RunInput) (*api.RunOutput, error) {
```
Comment (Stebalien) [inline pkg/runner/cluster_swarm.go]: No context?

File: pkg/runner/cluster_swarm.go [hunk 14]
```diff
@@ -1 +1,318 @@
...
+				switch status[s]++; s {
+				case swarm.TaskStateShutdown, swarm.TaskStateComplete, swarm.TaskStateFailed, swarm.TaskStateRejected:
+					finished++
+				}
+			}
+
+			logging.S().Infow("task status", "status", status)
+
+			if finished == replicas {
+				rc.Close()
```
Comment (Stebalien) [inline pkg/runner/cluster_swarm.go]: defer close?

File: pkg/runner/cluster_swarm.go [hunk 15]
```diff
@@ -1 +1,318 @@
...
+
+		tick := time.NewTicker(2 * time.Second)
+		defer tick.Stop()
+
+		for range tick.C {
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			tasks, err := cli.TaskList(ctx, types.TaskListOptions{
+				Filters: filters.NewArgs(filters.Arg("service", serviceID)),
+			})
+			cancel()
```
Comment (Stebalien) [inline pkg/runner/cluster_swarm.go]: someone's going to use this context. We should probably scope `ctx` so nobody does.

File: cmd/run.go (hunk 7 above)
Comment (raulk) [reply cmd/run.go]: 👍 That simplifies things.

File: cmd/run.go (hunk 8 above)
Comment (raulk) [reply cmd/run.go]: This library uses the full names of flags to obtain their values, in this case `no-build` (POSIX naming scheme).

File: pkg/aws/ecr.go (hunk 9 above)
Comment (raulk) [reply pkg/aws/ecr.go]: Yep.

File: pkg/build/golang/docker.go (hunk 10 above)
Comment (raulk) [reply pkg/build/golang/docker.go]: I was putting in some boilerplate for future registry types, and experimenting with idioms for switching without nesting large code blocks. But OK :P

File: pkg/engine/engine.go (hunk 11 above)
Comment (raulk) [reply pkg/engine/engine.go]: I guess so ;-)

File: pkg/engine/engine.go (hunk 12 above)
Comment (raulk) [reply pkg/engine/engine.go]: Yes, I thought I had changed that!

File: pkg/runner/cluster_swarm.go (hunk 14 above)
Comment (raulk) [reply pkg/runner/cluster_swarm.go]: I just wanted it to make it more evident, as the whole purpose of this goroutine is to close that `io.ReadCloser`. People tend to gloss over defers.

File: pkg/runner/cluster_swarm.go (hunk 13 above)
Comment (raulk) [reply pkg/runner/cluster_swarm.go]: Not for now. But we should use contexts: https://github.com/ipfs/testground/issues/131.

**Comment by daviddias** [comment]
Only missing a @Stebalien 👍🏽 for the merge? :)

File: pkg/runner/local_docker.go [hunk 16]
```diff
@@ -178,10 +177,14 @@ func (*LocalDockerRunner) Run(input *api.RunInput) (*api.RunOutput, error) {
 		containers = append(containers, res.ID)
 	}
 
+	if !cfg.KeepContainers {
+		defer deleteContainers(cli, log, containers)
```
Comment (daviddias) [inline pkg/runner/local_docker.go]: Go is unhappy here - https://github.com/ipfs/testground/pull/126#pullrequestreview-313209888

```suggestion
		defer func () {
  		err := deleteContainers(cli, log, containers)
  		if err != nil {
  		   // abort here?
  		}
		}
```
Or is there other best practice to silence this lint error?

Comment (Stebalien) [reply pkg/runner/local_docker.go]: In this case, we probably want to log.

Comment (Stebalien) [reply pkg/runner/local_docker.go]: (fixed)

**Pull request merged**